kpatch-build/lookup.c

```diff
@@ -363,8 +364,9 @@ static void symvers_read(struct lookup_table *table, char *path)
 		objname = make_modname(mod);
 
 		table->exp_syms[i].name = symname;
-		table->exp_syms[i].objname = objname;
+		table->exp_syms[i++].objname = objname;
 	}
+	table->exp_nr = i;
```

Comment:
Looking at the loop starting at line 321, I don't see what this changes. The value of table->exp_nr should be the same as before since both loops perform the same number of iterations.

Can you provide more detail on how to trigger the issue you are observing?

Reply:
AFAICT if `if (!name || !mod)` holds we might have a situation where there are some NULL symbols in table->exp_syms and none of the users of for_each_exp_symbol check for non-NULL causing a SIGSEGV.

This will fix the issue but maybe we shouldn't over-allocate either.

Reply:
Yes, previous code will leave some NULL symbols in table->exp_syms, so users of for_each_exp_symbol will trigger SIGSEGV when we test.
This patch will only inc i for each non-NULL symbol, so table->exp_nr will be the correct number.
And we can also avoid over-allocate in the first place.
I think both way is correct.

Reply:
Ah, you're right, I missed that.
In which case that that happen? Blank line?
How come the issue was not hit before?

Reply:
@julien-thierry Well, I actually don't know much about the inside logic or in which case the condition "if (!name || !mod)" will be true. Maybe in most cases for_each_exp_symbol break out before reach NULL symbols ?

Reply:
@chenIs I don't ever remember seeing NULL symbols, which kernel version had that?  Or if the problem is just theoretical, we could just do a fatal ERROR on a NULL symbol.

Otherwise, if it's a real problem, the patch looks ok, though the earlier uses of table->exp_nr in that function should just be replaced with a local variable.  We don't need to worry about overallocation.

Reply:
@chengmingzhou, sorry, github wouldn't autocomplete your name for some reason, and I entered in the wrong name twice.  I finally got it right :-)

Reply:
@jpoimboe Hi, we encountered this problem when we make a livepatch hotpatch for mm/mempolicy.c of kernel 4.14 using a bugfix source patch from our inside team.
I don't know how NULL symbols came out, create-diff-object just crash with SIGSEGV, so I use gdb to find this bug.
But today I pulled a new 4.14 kernel and kpatch-build again, this SIGSEGV didn't show up...
If NULL symbols should not be there theoretically, I think a fatal ERROR is ok too :-)
Thanks!

Reply:
@jpoimboe Hi, I replace the use of table->exp_nr with a local variable in that function.  Are there plans for this PR to be merged?

Thanks.

Reply:
@chengmingzhou If we don't expect NULL symbols to be possible, I think I'd rather go with an ERROR when a NULL symbol is encountered.

Reply:
@zhouchengming1 : any chance of attaching the symvers file that causes the crash?  Also, a gdb backtrace would be helpful (even in the commit msg) to better understand what conditions may lead to the crash.